src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProject.cs

```diff
@@ -46,24 +49,25 @@ internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
     /// <param name="existingReverseReferencesMap">The <see cref="_lazyReverseReferencesMap"/> prior to the removal.
     /// This map serves as a hint to the removal process; i.e. it is assumed correct if it contains data, but may be
     /// omitted without impacting correctness.</param>
-    /// <param name="removedProjectId">The ID of the project which is being removed.</param>
+    /// <param name="removedProjectIds">IDs of projects which are being removed.</param>
     /// <returns>The <see cref="_referencesMap"/> for the project dependency graph once the project is removed.</returns>
     private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReferencesMapForRemovedProject(
         ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
         ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
-        ProjectId removedProjectId)
+        ArrayBuilder<ProjectId> removedProjectIds)
     {
         var builder = existingForwardReferencesMap.ToBuilder();
 
-        if (existingReverseReferencesMap is object)
+        if (existingReverseReferencesMap is not null)
         {
-            // We know all the projects directly referencing 'projectId', so remove 'projectId' from the set of
-            // references in each of those cases directly.
-            if (existingReverseReferencesMap.TryGetValue(removedProjectId, out var referencingProjects))
+            foreach (var removedProjectId in removedProjectIds)
```

Comment:
I worry this might be potentially a bit slow if we're removing a lot of projects at once if the maps were to be full. The extreme case being if we're removing all projects, we'll call builder.MultiRemove n^2 times. There might be an improvement here just to build a list of all the projects that need to be updated, and then have MultiRemove take a hashset so we're creating fewer intermediate sets.

I'm 100% OK deferring this until we actually see a perf problem, or track it for futher investigation if it is one.

Reply:
As mentioned in the description, currently these arrays are size 1. *If* we are able to batch these removals, then it will definitely be a process of determining what the costs are in the batch and seeing how that can be addressed.

Reply:
> As mentioned in the description, currently these arrays are size 1. If we are able to batch these removals, then it will definitely be a process of determining what the costs are in the batch and seeing how that can be addressed.

My assumption is by merging this work we're considering that optimization a foregone conclusion, otherwise why merge this?

Reply:
But if the answer is "we want to make this work, then we'll do batching, then we'll see how to optimize this more in real life" I'm OK with that. Maybe leave a TODO comment so it's clear that was the intent.